File: src/vibe_kanban_clone/services/tasks.py
"""Task service."""

import structlog
from sqlalchemy import select, text
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from vibe_kanban_clone.models.agent import Agent
from vibe_kanban_clone.models.column import Column
from vibe_kanban_clone.models.task import Task
from vibe_kanban_clone.schemas.task import TaskCreate, TaskMove, TaskUpdate

logger = structlog.get_logger()


async def create_task(session: AsyncSession, data: TaskCreate) -> Task:
    """Create a new task."""
    task = Task(
        project_id=data.project_id,
        column_id=data.column_id,
        title=data.title,
        description=data.description,
        priority=data.priority,
        assigned_agent_id=data.assigned_agent_id,
    )
    session.add(task)

    col_result = await session.execute(select(Column).where(Column.id == task.column_id))
    column = col_result.scalar_one_or_none()
    if column is None or column.project_id != task.project_id:
        raise ValueError("Column does not belong to project")

    await session.commit()
    await session.refresh(task)
    logger.info("task_created", task_id=task.id)
    return task


async def get_task(session: AsyncSession, task_id: int) -> Task | None:
    """Get a task by ID with eagerly loaded comments."""
    result = await session.execute(
        select(Task).options(selectinload(Task.comments)).where(Task.id == task_id)
    )
    return result.scalar_one_or_none()


async def list_tasks_by_project(
    session: AsyncSession, project_id: int, limit: int = 50, offset: int = 0
) -> list[Task]:
    """List all tasks in a project."""
    result = await session.execute(
        select(Task).where(Task.project_id == project_id).limit(limit).offset(offset)
    )
    return list(result.scalars().all())


async def list_tasks_filtered(
    session: AsyncSession,
    project_id: int,
    status: str | None = None,
    agent_id: int | None = None,
    priority_gte: int | None = None,
) -> list[Task]:
    """List tasks with optional filters."""
    stmt = select(Task).where(Task.project_id == project_id)
    if status is not None:
        stmt = stmt.outerjoin(Column, Task.column_id == Column.id)
        if status == "todo":
            stmt = stmt.where(Task.claimed_at.is_(None))
        elif status == "in_progress":
            stmt = stmt.where(Task.claimed_at.is_not(None), Column.is_terminal == False)  # noqa: E712
        elif status == "done":
            stmt = stmt.where(Column.is_terminal == True)  # noqa: E712
    if agent_id is not None:
        stmt = stmt.where(Task.assigned_agent_id == agent_id)
    if priority_gte is not None:
        stmt = stmt.where(Task.priority >= priority_gte)
    stmt = stmt.order_by(Task.priority.asc(), Task.created_at.asc())
    result = await session.execute(stmt)
    return list(result.scalars().all())


async def update_task(session: AsyncSession, task: Task, data: TaskUpdate) -> Task:
    """Update a task, checking expected_version for optimistic locking."""
    if task.version != data.expected_version:
        raise ValueError("version mismatch")

    if data.title is not None:
        task.title = data.title
    if data.description is not None:
        task.description = data.description
    if data.priority is not None:
        task.priority = data.priority
    if data.result is not None:
        task.result = data.result
    if data.assigned_agent_id is not None:
        agent_result = await session.execute(
            select(Agent).where(Agent.id == data.assigned_agent_id)
        )
        agent = agent_result.scalar_one_or_none()
        if agent is None:
            raise ValueError("Agent not found")
        task.assigned_agent_id = data.assigned_agent_id

    task.version += 1
    await session.commit()
    await session.refresh(task)
    logger.info("task_updated", task_id=task.id, version=task.version)
    return task


async def move_task(session: AsyncSession, task: Task, data: TaskMove) -> Task:
    """Move a task to another column, checking expected_version."""
    if task.version != data.expected_version:
        raise ValueError("version mismatch")

    col_result = await session.execute(select(Column).where(Column.id == data.column_id))
    column = col_result.scalar_one_or_none()
    if column is None or column.project_id != task.project_id:
        raise ValueError("Column does not belong to project")

    task.column_id = data.column_id
    task.version += 1
    await session.commit()
    await session.refresh(task)
    logger.info("task_moved", task_id=task.id, column_id=task.column_id)
    return task


async def delete_task(session: AsyncSession, task: Task) -> None:
    """Delete a task."""
    await session.delete(task)
    await session.commit()
    logger.info("task_deleted", task_id=task.id)


async def assign_agent_to_task(session: AsyncSession, task: Task, agent_id: int) -> Task:
    """Assign an agent to a task."""
    task.assigned_agent_id = agent_id
    await session.commit()
    await session.refresh(task)
    logger.info("task_assigned", task_id=task.id, agent_id=agent_id)
    return task


async def unassign_agent(session: AsyncSession, task: Task) -> Task:
    """Unassign the agent from a task."""
    task.assigned_agent_id = None
    await session.commit()
    await session.refresh(task)
    logger.info("task_unassigned", task_id=task.id)
    return task


async def claim_next_task(session: AsyncSession, agent_id: int, project_id: int) -> Task | None:
    """Atomically claim the next available task for an agent."""
    sql = text(
        """
        UPDATE tasks
        SET
            column_id = (
                SELECT c2.id
                FROM columns c2
                WHERE c2.project_id = :project_id
                AND c2.position > (
                    SELECT c.position FROM columns c WHERE c.id = tasks.column_id
                )
                ORDER BY c2.position ASC
                LIMIT 1
            ),
            claimed_at = CURRENT_TIMESTAMP,
            assigned_agent_id = :agent_id,
            version = version + 1
        WHERE id = (
            SELECT t.id
            FROM tasks t
            JOIN columns c ON t.column_id = c.id
            WHERE t.project_id = :project_id
            AND t.claimed_at IS NULL
            AND c.is_terminal = FALSE
            AND EXISTS (
                SELECT 1 FROM columns c2
                WHERE c2.project_id = :project_id
                AND c2.position > c.position
            )
            ORDER BY t.priority ASC, t.created_at ASC
            LIMIT 1
        )
        RETURNING *
        """
    )
    result = await session.execute(sql, {"project_id": project_id, "agent_id": agent_id})
    row = result.mappings().fetchone()
    await session.commit()
    if row is None:
        return None
    task = await session.get(Task, row["id"])
    if task is None:
        return None

    logger.info("task_claimed", task_id=task.id, agent_id=agent_id)
    return task


async def complete_task(
    session: AsyncSession, task: Task, result: str, terminal_column_id: int | None = None
) -> Task:
    """Mark a task as complete by writing the result and moving it to the terminal column."""
    if terminal_column_id is None:
        col_result = await session.execute(
            select(Column)
            .where(Column.project_id == task.project_id, Column.is_terminal == True)  # noqa: E712
            .order_by(Column.position.asc())
            .limit(1)
        )
        terminal_col = col_result.scalar_one_or_none()
        if terminal_col is None:
            raise ValueError("no terminal column")
        terminal_column_id = terminal_col.id

    task.result = result
    task.column_id = terminal_column_id
    task.version += 1
    await session.commit()
    await session.refresh(task)
    logger.info("task_completed", task_id=task.id)
    return task

File: server/src/services/tasks.ts
import { Prisma } from "@prisma/client";
import { prisma } from "../prisma/client.js";
import { logger } from "../lib/logger.js";
import type { TaskCreate, TaskMove, TaskUpdate } from "../schemas/task.js";

export async function createTask(data: TaskCreate) {
  const task = await prisma.task.create({
    data: {
      project_id: data.project_id,
      column_id: data.column_id,
      title: data.title,
      description: data.description,
      priority: data.priority,
      assigned_agent_id: data.assigned_agent_id,
    },
  });

  if (task.column_id !== null) {
    const column = await prisma.column.findUnique({
      where: { id: task.column_id },
    });
    if (!column || column.project_id !== task.project_id) {
      throw new Error("Column does not belong to project");
    }
  }

  logger.info({ task_id: task.id }, "task_created");
  return task;
}

export async function getTask(taskId: number) {
  return prisma.task.findUnique({
    where: { id: taskId },
    include: { comments: true },
  });
}

export async function listTasksByProject(projectId: number, limit = 50, offset = 0) {
  return prisma.task.findMany({
    where: { project_id: projectId },
    take: limit,
    skip: offset,
  });
}

export async function listTasksFiltered(
  projectId: number,
  status?: string,
  agentId?: number,
  priorityGte?: number
) {
  const conditions: string[] = [`t.project_id = ${projectId}`];

  if (status === "todo") {
    conditions.push("t.claimed_at IS NULL");
  } else if (status === "in_progress") {
    conditions.push("t.claimed_at IS NOT NULL");
    conditions.push("c.is_terminal = 0");
  } else if (status === "done") {
    conditions.push("c.is_terminal = 1");
  }

  if (agentId !== undefined) {
    conditions.push(`t.assigned_agent_id = ${agentId}`);
  }
  if (priorityGte !== undefined) {
    conditions.push(`t.priority >= ${priorityGte}`);
  }

  const whereClause = conditions.join(" AND ");

  const rows = await prisma.$queryRawUnsafe<
    Array<{
      id: number;
      project_id: number;
      column_id: number | null;
      title: string;
      description: string | null;
      priority: number;
      result: string | null;
      assigned_agent_id: number | null;
      version: number;
      claimed_at: Date | null;
      created_at: Date;
      updated_at: Date;
    }>
  >(`
    SELECT t.* FROM tasks t
    LEFT JOIN columns c ON t.column_id = c.id
    WHERE ${whereClause}
    ORDER BY t.priority ASC, t.created_at ASC
  `);

  return rows;
}

export async function updateTask(taskId: number, data: TaskUpdate) {
  const task = await prisma.task.findUnique({ where: { id: taskId } });
  if (!task) throw new Error("Task not found");
  if (data.expected_version !== undefined && task.version !== data.expected_version) {
    throw new Error("version mismatch");
  }

  const updateData: Prisma.TaskUpdateInput = {};
  if (data.title !== undefined) updateData.title = data.title;
  if (data.description !== undefined) updateData.description = data.description;
  if (data.priority !== undefined) updateData.priority = data.priority;
  if (data.result !== undefined) updateData.result = data.result;
  if (data.assigned_agent_id !== undefined) {
    if (data.assigned_agent_id !== null) {
      const agent = await prisma.agent.findUnique({ where: { id: data.assigned_agent_id } });
      if (!agent) throw new Error("Agent not found");
      updateData.assignedAgent = { connect: { id: data.assigned_agent_id } };
    } else {
      updateData.assignedAgent = { disconnect: true };
    }
  }

  const updated = await prisma.task.update({
    where: { id: taskId },
    data: { ...updateData, version: { increment: 1 } },
  });
  logger.info({ task_id: updated.id, version: updated.version }, "task_updated");
  return updated;
}

export async function moveTask(taskId: number, data: TaskMove) {
  const task = await prisma.task.findUnique({ where: { id: taskId } });
  if (!task) throw new Error("Task not found");
  if (data.expected_version !== undefined && task.version !== data.expected_version) {
    throw new Error("version mismatch");
  }

  const column = await prisma.column.findUnique({ where: { id: data.column_id } });
  if (!column || column.project_id !== task.project_id) {
    throw new Error("Column does not belong to project");
  }

  const updated = await prisma.task.update({
    where: { id: taskId },
    data: { column_id: data.column_id, version: { increment: 1 } },
  });
  logger.info({ task_id: updated.id, column_id: updated.column_id }, "task_moved");
  return updated;
}

export async function deleteTask(taskId: number) {
  await prisma.task.delete({ where: { id: taskId } });
  logger.info({ task_id: taskId }, "task_deleted");
}

export async function assignAgentToTask(taskId: number, agentId: number) {
  const updated = await prisma.task.update({
    where: { id: taskId },
    data: { assigned_agent_id: agentId },
  });
  logger.info({ task_id: taskId, agent_id: agentId }, "task_assigned");
  return updated;
}

export async function unassignAgent(taskId: number) {
  const updated = await prisma.task.update({
    where: { id: taskId },
    data: { assigned_agent_id: null },
  });
  logger.info({ task_id: taskId }, "task_unassigned");
  return updated;
}

export async function claimNextTask(agentId: number, projectId: number) {
  const rows = await prisma.$queryRawUnsafe<
    Array<{
      id: number;
      project_id: number;
      column_id: number | null;
      title: string;
      description: string | null;
      priority: number;
      result: string | null;
      assigned_agent_id: number | null;
      version: number;
      claimed_at: Date | null;
      created_at: Date;
      updated_at: Date;
    }>
  >(`
    UPDATE tasks
    SET
      column_id = (
        SELECT c2.id
        FROM columns c2
        WHERE c2.project_id = ${projectId}
        AND c2.position > (
          SELECT c.position FROM columns c WHERE c.id = tasks.column_id
        )
        ORDER BY c2.position ASC
        LIMIT 1
      ),
      claimed_at = CURRENT_TIMESTAMP,
      assigned_agent_id = ${agentId},
      version = version + 1
    WHERE id = (
      SELECT t.id
      FROM tasks t
      JOIN columns c ON t.column_id = c.id
      WHERE t.project_id = ${projectId}
      AND t.claimed_at IS NULL
      AND c.is_terminal = FALSE
      AND EXISTS (
        SELECT 1 FROM columns c2
        WHERE c2.project_id = ${projectId}
        AND c2.position > c.position
      )
      ORDER BY t.priority ASC, t.created_at ASC
      LIMIT 1
    )
    RETURNING *
  `);

  if (rows.length === 0) return null;
  const row = rows[0];
  logger.info({ task_id: row.id, agent_id: agentId }, "task_claimed");

  return prisma.task.findUnique({ where: { id: row.id } });
}

export async function completeTask(taskId: number, result: string) {
  const task = await prisma.task.findUnique({ where: { id: taskId } });
  if (!task) throw new Error("Task not found");

  const terminalCol = await prisma.column.findFirst({
    where: { project_id: task.project_id, is_terminal: true },
    orderBy: { position: "asc" },
  });
  if (!terminalCol) throw new Error("no terminal column");

  const updated = await prisma.task.update({
    where: { id: taskId },
    data: { result, column_id: terminalCol.id, version: { increment: 1 } },
  });
  logger.info({ task_id: updated.id }, "task_completed");
  return updated;
}
